replication-api-impl/src/main/java/com/connexta/replication/api/impl/worker/Worker.java

```diff
@@ -0,0 +1,259 @@
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Handles a single transfer between a local and remote Site based on the tasks received from the
+ * queue. This worker will respond to thread interruptions.
+ */
+public class Worker extends Thread {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(Worker.class);
+
+  private final AtomicBoolean processing = new AtomicBoolean(false);
+
+  private volatile boolean running = true;
+
+  private final NodeAdapter local;
+
+  private final NodeAdapter destination;
+
+  private final Queue queue;
+
+  /**
+   * Creates a new worker which will monitor the provided queue and perform operations between the
+   * provided local and destination sites.
+   *
+   * Does not begin monitoring of the queue until {@link #run()} is called.
+   *
+   * @param queue the queue to pull tasks from; corresponding queue for the destination site
+   * @param local the local site adapter
+   * @param destination the destination site adapter
+   */
+  public Worker(Queue queue, NodeAdapter local, NodeAdapter destination) {
+    this.queue = queue;
+    this.local = local;
+    this.destination = destination;
+  }
+
+  /**
+   * Begin monitoring a queue for tasks. The handler of this worker is responsible for halting this
+   * thread by calling interrupt on this worker's thread, finishing the transfer of the current task
+   * if necessary.
+   */
+  @Override
+  public void run() {
+    while (running || !Thread.currentThread().isInterrupted()) {
+      try {
+        Task task = queue.take();
+        this.processing.set(true);
+        if (supported(task)) {
+          this.handle(task);
+        }
+        this.processing.set(false);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
+
+  /**
+   * Set the running flag to false and wait for the thread to die. If an item is currently in the
+   * process of transferring, it will finish.
+   */
+  public void cancel() {
+    this.running = false;
+    try {
+      Thread.currentThread().join();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+    }
+  }
+
+  /**
+   * @return {@code true} if this worker is transferring an item, otherwise {@code false}
+   */
+  public boolean isProcessing() {
+    return processing.get();
+  }
+
+  private boolean supported(Task task) throws InterruptedException {
+    if (!OperationType.UNKNOWN.equals(task.getOperation())) {
+      task.fail(ErrorCode.UNSUPPORTED_ERROR,
+          String.format("Operation task %s is not supported by this worker", task));
+      return false;
+    }
+    return true;
+  }
+
+  private void handle(Task task) throws InterruptedException {
+    boolean proceed = checkAdapterAvailabilities(task);
+    if (!proceed) {
+      return;
+    }
+
+    List<DDFMetadataInfo> infos = task.metadatas()
+        .filter(t -> t instanceof DDFMetadataInfo)
```

Comment:
```suggestion
        .filter(DDFMetadataInfo.class::isInstance)
```
❓ How will that work later when dealing with Ion? If we don't support it yet (which I know we don't) then we should make sure to filter these tasks out as soon as possible and let someone else handle them (like a newer version). This check shuld be added to your `supported()` method.

Reply:
Added the unknown, size and type checks to the supported method and replaced the failures with unlocks.